### PR TITLE
[Issue-2536] Add support to orders batch update

### PR DESCRIPTION
**Pull request by rodrigorods**
Closes [#2536](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/2536)
Initial P2 discussion [here](https://woomobilep2.wordpress.com/2022/10/14/trial-task-batch-orders-support/) 

This PR enables user to perform multiple actions, Create, Update and/or Delete Orders, to be performed on same request, to a maximum of 100.

Done - 
1. Error handling. ✅ [commit](b1bb2bd218d52926084b0263b8697033dbf50866)
2. UI of test app that allow creation of several Orders and Updates ✅ [commit](da52b97e2e116cad7ae1e5bde2945e1db72a616d)

First fetched orders | Fetching orders after batch update
--- | --- |
Screenshot_20221017_003203 | Screenshot_20221017_003250

### CODE REVIEW

File: example/src/main/java/org/wordpress/android/fluxc/example/ui/orders/WooOrdersFragment.kt [hunk 1]
```diff
@@ -597,6 +597,88 @@ class WooOrdersFragment : StoreSelectingFragment(), WCAddOrderShipmentTrackingDi
                 }
             }
         }
+
+        update_order_batch.setOnClickListener {
```
Comment (atorresveiga) [inline example/src/main/java/org/wordpress/android/fluxc/example/ui/orders/WooOrdersFragment.kt:570]: Nice! Great work adding the update order batch request to the FluxC example

File: example/src/main/java/org/wordpress/android/fluxc/example/ui/orders/WooOrdersFragment.kt [hunk 2]
```diff
@@ -597,6 +597,88 @@ class WooOrdersFragment : StoreSelectingFragment(), WCAddOrderShipmentTrackingDi
...
+                    val products = showSingleLineDialog(
+                        activity = requireActivity(),
+                        message = "Please type a comma separated list of product IDs",
+                        isNumeric = false
+                    )?.split(",")?.map { it.toLongOrNull() }
+
+                    if (products == null || products.any { it == null }) {
+                        prependToLog("Error while parsing the entered product IDs")
+                        return@launch
+                    }
```
Comment (atorresveiga) [inline example/src/main/java/org/wordpress/android/fluxc/example/ui/orders/WooOrdersFragment.kt]: I noted that here the `create` request is mandatory. Taking into account that `create`, `update` and `delete` requests are optional in the order batch request. Does it make any sense to make this request also optional? This way, it will be easy to test only one of them.

File: example/src/main/java/org/wordpress/android/fluxc/example/ui/orders/WooOrdersFragment.kt [hunk 3]
```diff
@@ -597,6 +597,88 @@ class WooOrdersFragment : StoreSelectingFragment(), WCAddOrderShipmentTrackingDi
...
+                        shippingAddress = shippingAddress,
+                        billingAddress = billingAddress,
+                        customerNote = customerNote
+                    )
+
+                    val result = orderUpdateStore.updateOrdersBatch(
+                        site,
+                        createRequest = mutableListOf<UpdateOrderRequest>().apply {
+                            repeat(duplicatesCount) { this.add(orderToCreate) }
+                        },
```
Comment (atorresveiga) [inline example/src/main/java/org/wordpress/android/fluxc/example/ui/orders/WooOrdersFragment.kt]: np: I understand that here your goal is to create a list initialized with the same `orderToCreate` object `duplicatesCount` times. I feel that using the list constructor would be better suited for your goal here. 
Something like :
```List(duplicatesCount) { orderToCreate } ```.  WDYT?

Is the `mutableListOf` function used instead of the `listOf` function for any particular reason here?

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClient.kt [hunk 4]
```diff
@@ -815,6 +862,13 @@ class OrderRestClient @Inject constructor(
         }
     }
 
+    private fun buildOrderEntity(
```
Comment (atorresveiga) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClient.kt:865]: Nice!

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrdersBatchDto.kt [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+object OrdersBatchUpdateRequest {
+    fun buildBody(
+        createRequest: List<Map<String, Any>>,
+        updateRequest: List<Map<String, Any>>,
+        deleteRequest: List<Long>
+    ) = mutableMapOf<String, Any>().apply {
+        put("create", createRequest)
+        put("update", updateRequest)
+        put("delete", deleteRequest)
```
Comment (atorresveiga) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrdersBatchDto.kt]: Maybe the request could be removed from the body when we know the list is empty so we ensure that only necessary data is sent. WDYT?

Is this map mutable for any particular reason?

Maybe the [buildMap](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/build-map.html) function could be handy here.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/OrderUpdateStore.kt [hunk 6]
```diff
@@ -231,6 +232,41 @@ class OrderUpdateStore @Inject internal constructor(
         }
     }
 
+    suspend fun updateOrdersBatch(
+        site: SiteModel,
+        createRequest: List<UpdateOrderRequest> = emptyList(),
+        updateRequest: List<UpdateOrderRequest> = emptyList(),
+        deleteRequest: List<Long> = emptyList()
```
Comment (atorresveiga) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/OrderUpdateStore.kt:239]: I'm curious about your decision-making process here. Why did you favor the default to `emptyList()` approach here rather than the default  to `null` one?

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/OrderUpdateStore.kt [hunk 7]
```diff
@@ -231,6 +232,41 @@ class OrderUpdateStore @Inject internal constructor(
...
+
+                if (createRequest.isNotEmpty()) {
+                    model.createdEntities.map { ordersDao.insertOrUpdateOrder(it) }
+                    requestListRefresh(site)
+                }
+                if (updateRequest.isNotEmpty()) {
+                    model.updatedEntities.map { ordersDao.insertOrUpdateOrder(it) }
+                }
+                if (deleteRequest.isNotEmpty()) {
+                    deleteRequest.map { ordersDao.deleteOrder(site.localId(), it) }
```
Comment (atorresveiga) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/OrderUpdateStore.kt]: Shouldn't the orders be deleted only if they are present in the response (model.deletedEntities)?

File: plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClientTest.kt [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    @Test
+    fun `updateOrdersBatch should call gson builder request with expected params`(): Unit = runBlocking {
+        val createRequest = listOf(
+            buildOrderRequest("create1"), buildOrderRequest("create2")
+        )
+        val updateRequest = listOf(
+            buildOrderRequest("update1"), buildOrderRequest("update2")
+        )
+        val deleteRequest = listOf(
+            1.toLong(), 2.toLong(), 3.toLong()
```
Comment (atorresveiga) [inline plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClientTest.kt]: np: A long value could be represented in Kotlin as `1L`  instead of `1.toLong()`. Using the `1L` representation IMO is short and more Kotlin-ish. Do you agree with me?

File: plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/store/OrderUpdateStoreTest.kt [hunk 9]
```diff
@@ -520,6 +522,137 @@ class OrderUpdateStoreTest {
...
+    @Test
+    fun `should delete local copy of order when bulky delete request succeeds`(): Unit = runBlocking {
+        setUp {
+            orderRestClient = mock {
+                onBlocking {
+                    updateOrdersBatch(
+                        site = site,
+                        createRequest = emptyList(),
+                        updateRequest = emptyList(),
+                        deleteRequest = listOf(1.toLong(), 2.toLong())
```
Comment (atorresveiga) [inline plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/store/OrderUpdateStoreTest.kt]: [Same comments as here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2539/files#r998499561)

Would not be better to maintain the same ids in the requests and the responses? I see that the function is being called with ids `1` and `2`, but in the mock response, the ids are set as `123` and `456`

Maybe we can extract the list of order ids to a value and use it on the mock response when calling the function. WDYT?

File: plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/store/OrderUpdateStoreTest.kt [hunk 10]
```diff
@@ -520,6 +522,137 @@ class OrderUpdateStoreTest {
...
+                    )
+                }.doReturn(
+                    WooPayload(
+                        OrdersDatabaseBatch(
+                            createdEntities = emptyList(),
+                            updatedEntities = emptyList(),
+                            deletedEntities = listOf(
+                                OrderEntity(orderId = 123.toLong(), localSiteId = LocalId(1)),
+                                OrderEntity(orderId = 456.toLong(), localSiteId = LocalId(2))
```
Comment (atorresveiga) [inline plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/store/OrderUpdateStoreTest.kt]: It would be great if we also test edge cases. For example: 
What would happen if we call the batch delete function with ids `1,2,3` but the response only includes ids `1,3`?

File: plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClientTest.kt [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+import org.wordpress.android.fluxc.model.order.OrderAddress.Shipping
+import org.wordpress.android.fluxc.model.order.UpdateOrderRequest
+import org.wordpress.android.fluxc.network.UserAgent
+import org.wordpress.android.fluxc.network.rest.wpcom.auth.AccessToken
+import org.wordpress.android.fluxc.network.rest.wpcom.jetpacktunnel.JetpackTunnelGsonRequestBuilder
+import org.wordpress.android.fluxc.network.rest.wpcom.jetpacktunnel.JetpackTunnelGsonRequestBuilder.JetpackResponse.JetpackSuccess
+import org.wordpress.android.fluxc.network.rest.wpcom.wc.WooPayload
+
+@RunWith(MockitoJUnitRunner::class)
+class OrderRestClientTest {
```
Comment (atorresveiga) [inline plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClientTest.kt:32]: Nice! Thanks for adding unit tests to this class.

I believe we could also include a unit test for the failure case as on [LeaderboardsRestClientTest](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/da52b97e2e116cad7ae1e5bde2945e1db72a616d/example/src/test/java/org/wordpress/android/fluxc/wc/leaderboards/LeaderboardsRestClientTest.kt). WDYT?

**Comment by atorresveiga** [comment]
Great work Rodrigo! Thanks for adding more tests to the code and the batch update request to the FluxC example.
I left a few comments, and I'm looking forward to hearing your thoughts on those before merging the code.🙂

File: plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClientTest.kt [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        lastName = "lastName",
+        company = "company",
+        address1 = "address1",
+        address2 = "address2",
+        city = "city",
+        state = "state",
+        postcode = "postcode",
+        country = "countryCode",
+        phone = "phone"
+    )
```
Comment (atorresveiga) [inline plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClientTest.kt]: I wonder if we really need all this information. I don't see that we are taking advantage of this information in the test

Comment (rodrigorods) [reply plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClientTest.kt]: The idea was get to get the code ready to check the generated body with mockito. I did not wanted to make big changes on something unrelated to my code, and to get this done I would have to move function `UpdateOrderRequest.toNetworkRequest`. 

I will create a class `OrderRequestBodyBuilder ` and move all body construction code to it.
this will allow me to test the `toNetworkRequest()` code and check the body param at OrderRestClientTest.

Ex:
```
object OrderRequestBodyBuilder {
    fun build() { ... } // This would be the older toNetworkRequest() fun
    fun buildFromOrdersBatch() { ... }
}
```

Would you say that this is a good aproach?

File: plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClientTest.kt (hunk 11 above)
Comment (rodrigorods) [reply plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClientTest.kt:32]: Sure (: I will add tests for the error scenarios.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/OrderUpdateStore.kt (hunk 7 above)
Comment (rodrigorods) [reply plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/OrderUpdateStore.kt]: Yes, it makes more sense to delete only whats is available at the response. I will work on that.

File: plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClientTest.kt (hunk 12 above)
Comment (atorresveiga) [reply plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClientTest.kt]: My point above was that if substitute:
```
   val response = orderRestClient.updateOrdersBatch(
        site = site,
        createRequest = createRequest,
        updateRequest = updateRequest,
        deleteRequest = deleteRequest
    )
```
With 
```
   val response = orderRestClient.updateOrdersBatch(
        site = site,
        createRequest = emptyList(),
        updateRequest = emptyList(),
        deleteRequest = emptyList()
    )
```
The test will still pass, so maybe no need for the extra info.

After a second look, I don't think the test's main goal of ensuring **_updateOrdersBatch should call gson builder request with expected params_** can't be guaranteed. The test is checking that the response is a success and not checking the parameters used to call the gson builder. One suggestion here could be to use the `verify` function of mockito.

With this change, the test could look like this:
```
@Test
    fun `updateOrdersBatch should call gson builder request with expected params`(): Unit =
        runBlocking {
            val idsToDelete = listOf(1L, 2L, 3L)
            val body = OrdersBatchUpdateRequest.buildBody(
                emptyList(),
                emptyList(),
                idsToDelete
            )

            whenever(
                jetpackTunnelGsonRequestBuilder.syncPostRequest(
                    eq(orderRestClient),
                    site = eq(site),
                    url = eq(WOOCOMMERCE.orders.batch.pathV3),
                    body = any(),
                    clazz = eq(OrdersBatchDto::class.java)
                )
            ).thenReturn(
                JetpackSuccess(OrdersBatchDto())
            )

            orderRestClient.updateOrdersBatch(
                site = site,
                createRequest = emptyList(),
                updateRequest = emptyList(),
                deleteRequest = idsToDelete
            )

            verify(jetpackTunnelGsonRequestBuilder, times(1))
                .syncPostRequest(
                    restClient = eq(orderRestClient),
                    site = eq(site),
                    url = eq(WOOCOMMERCE.orders.batch.pathV3),
                    body = eq(body),
                    clazz = eq(OrdersBatchDto::class.java)
                )
        }
```
With this version, we don't need to modify `UpdateOrderRequest.toNetworkRequest`, and if we change the parameters we use to call the function without modifying the expected response, the test will fail.

File: example/src/main/java/org/wordpress/android/fluxc/example/ui/orders/WooOrdersFragment.kt [hunk 13]
```diff
@@ -597,6 +566,105 @@ class WooOrdersFragment : StoreSelectingFragment(), WCAddOrderShipmentTrackingDi
...
+                        UpdateOrderRequest(
+                            id = it,
+                            customerNote = "$newCostumerNote"
+                        )
+                    } ?: emptyList()
+
+                    if (
+                        ordersToCreate.isEmpty() &&
+                        ordersToUpdate.isEmpty() &&
+                        ordersToUpdate.isEmpty()
```
Comment (atorresveiga) [inline example/src/main/java/org/wordpress/android/fluxc/example/ui/orders/WooOrdersFragment.kt]: Note that here we are missing the `ordersToDelete` list and checking twice for the `ordersToUpdate`

File: plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/store/OrderUpdateStoreTest.kt [hunk 14]
```diff
@@ -574,8 +582,8 @@ class OrderUpdateStoreTest {
                         OrdersDatabaseBatch(
                             createdEntities = emptyList(),
                             updatedEntities = listOf(
-                                OrderEntity(orderId = 123.toLong(), localSiteId = LocalId(1)),
-                                OrderEntity(orderId = 456.toLong(), localSiteId = LocalId(2))
+                                OrderEntity(orderId = 123L, localSiteId = LocalId(1)),
+                                OrderEntity(orderId = 456L, localSiteId = LocalId(2))
```
Comment (atorresveiga) [inline plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/store/OrderUpdateStoreTest.kt]: If we are using the `orderEntities` for the assertion below (line 600). Wouldn't make sense to use the same list in the mocked result?
Something like this:
```
.doReturn(
      WooPayload(
          OrdersDatabaseBatch(
              createdEntities = emptyList(),
              updatedEntities = orderEntities,
              deletedEntities = emptyList()
          )
      )
  )
```
This way, if someone needs to make changes to `orderEntities`, the tests won't be affected.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/OrderUpdateStore.kt [hunk 15]
```diff
@@ -260,7 +260,7 @@ class OrderUpdateStore @Inject internal constructor(
                     model.updatedEntities.map { ordersDao.insertOrUpdateOrder(it) }
                 }
                 if (deleteRequest.isNotEmpty()) {
-                    deleteRequest.map { ordersDao.deleteOrder(site.localId(), it) }
+                    model.deletedEntities.map { ordersDao.deleteOrder(it.localSiteId, it.orderId) }
```
Comment (atorresveiga) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/OrderUpdateStore.kt:263]: A suggestion for the future is to submit these changes in a separate commit, so it is easy for the reviewer to track those.

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrdersBatchDto.kt [hunk 16]
```diff
@@ -15,10 +15,14 @@ object OrdersBatchUpdateRequest {
...
         deleteRequest: List<Long>
-    ) = mutableMapOf<String, Any>().apply {
-        put("create", createRequest)
-        put("update", updateRequest)
-        put("delete", deleteRequest)
+    ) = buildMap {
+        putNotEmpty("create", createRequest)
+        putNotEmpty("update", updateRequest)
+        putNotEmpty("delete", deleteRequest)
+    }
```
Comment (atorresveiga) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrdersBatchDto.kt]: Nice!

File: plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrdersBatchDto.kt [hunk 17]
```diff
@@ -15,10 +15,14 @@ object OrdersBatchUpdateRequest {
...
+
+    private fun MutableMap<String, Any>.putNotEmpty(key: String, value: List<*>) {
+        if (value.isNotEmpty()) this[key] = value
     }
```
Comment (atorresveiga) [inline plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrdersBatchDto.kt]: Checking the codebase, I found that we have a very similar extension function [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/trunk/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/utils/MapExt.kt).
Maybe we could add this extension function to the same file with some small changes to make it more generic.
```
fun <K, V : List<*>> MutableMap<K, V>.putIfNotEmpty(key: K, value: V) {
  if (value.isNotEmpty()) this[key] = value
}
```
And make it available to everyone who may need it. WDYT?

File: plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrdersBatchUpdateRequestTest.kt [hunk 18]
```diff
@@ -0,0 +1,56 @@
+package org.wordpress.android.fluxc.network.rest.wpcom.wc.order
+
+import org.assertj.core.api.Assertions
+import org.junit.Test
+
+class OrdersBatchUpdateRequestTest {
```
Comment (atorresveiga) [inline plugins/woocommerce/src/test/java/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrdersBatchUpdateRequestTest.kt:6]: Nice!

**Comment by atorresveiga** [comment]
Closed until we resume the work on bulk update orders
